src/cls/IPM/ResourceProcessor/WebApplication.cls

```diff
@@ -0,0 +1,320 @@
+                For i=1:1:$Get(changes) {
+                    Write !,changes(i)
+                }
+            }
+            Set sc = ##class(Security.Applications).Modify(..Name,.properties)
+            $$$ThrowOnError(sc)
+            Write:pVerbose !,"Done."
+        } Else {
+            Write:pVerbose !,"No changes detected or made."
+        }
+    } Else {
+        Write:pVerbose !,"Creating Web Application ",..Name
+        Set sc = ##class(Security.Applications).Create(..Name,.properties)
+        If (pVerbose) {
+            Set key = ""
+            For {
+                Set key = $Order(properties(key),1,tValue)
+                Quit:key=""
+                Write !,key,": ",tValue
+            }
+        }
+        $$$ThrowOnError(sc)
+        Write:pVerbose !,"Done."
+    }
+}
+
+/// This removes an existing CSP application
+Method DeleteWebApp(pVerbose As %Boolean = 0) [ Internal ]
+{
+    New $Namespace
+    Set $Namespace = "%SYS"
+    If ##class(Security.Applications).Exists(..Name) {
+        $$$ThrowOnError(##class(Security.Applications).Delete(..Name))
+    }
+}
+
+/// Returns a unique name for this resource.
+/// Default implementation, leaving <var>pUniqueName</var> undefined, reverts to default behavior.
+/// An empty string indicates guaranteed uniqueness (useful for folders relative to module root)
+Method OnGetUniqueName(Output pUniqueName)
+{
+    If (..Name '= "") {
+        Set pUniqueName = ..Name
+    }
+}
+
+Method GetSourceControlInfo(Output pInfo As %IPM.ExtensionBase.SourceControl.ResourceInfo) As %Status
+{
+    Set pInfo = ##class(%IPM.ExtensionBase.SourceControl.ResourceInfo).%New()
+    If (..Name '= "") {
+        Set pInfo.SourceControlAware = 1
+        Set pInfo.ResourceType = "/CSP/"
+        Set pInfo.Prefix = ..Name
+        Set pInfo.RelativePath = ..ResourceReference.Name
+        Set pInfo.IsDirectory = 1
+    }
+    Quit $$$OK
+}
+
+/// Returns the path relative to the module root for item <var>pItemName</var> within this resource.
+Method OnItemRelativePath(pItemName As %String) As %String
+{
+    Quit ..ResourceReference.Name_$Piece(pItemName,..Name,2)
+}
+
+/// Check if the provided properties exist in Security.Applications.
+/// Returns a list of properties that do not exist.
+ClassMethod DoPropertiesExist(pPropList As %Library.List) As %Library.List [ Internal, Private ]
+{
+    New $Namespace
+    Set $Namespace = "%SYS"
+
+    Set badList = ""
+    Set ptr = 0
+    While $ListNext(pPropList,ptr,prop) {
+        If '$$$comMemberDefined("Security.Applications",$$$cCLASSproperty,prop) {
+            Set badList = badList _ $ListBuild(prop)
+        }
+    }
+    Return badList
+}
+
+Method ConstructRoleName(pDbResource As %String) As %String [ Internal, Private ]
+{
+    // Include the module name in the role name to avoid conflicts
+    // This role should only be managed by IPM itself
+    Return $ListToString($lb("IPM", "readonly", ..ResourceReference.Module.Name, pDbResource), ".")
```

Comment:
Max length of a role name is 64 characters so this could definitely go over the max length with long module names. For example, in HS we have a module named HealthShare.HP.OAuth2.ResourceServer so the role name "IPM.readonly.HealthShare.HP.OAuth2.ResourceServer.%DB_HSREGISTRY" has length 64 already which is the maxlen so a db/namespace name that is slightly longer will push it over the edge